### ISSUE TITLE
problem about Recurrence experiment in Web30k

### ISSUE THREAD

**Issue by stanpcf** [zhihu/SERank#1]
Hi, thx for your great job. but I have problem about recurrence experiment in web30k. I download the code, and change data to web30k, ndcg@5 of the last ckpt is only 0.409, which is 0.456 in paper.
I do not change serank.py. and the run_script is 
```bash
DATA=/path-to-web30k/Fold1

output_dir=outputs
rm -r $output_dir
mkdir $output_dir

python serank.py \
  --train_path=$DATA/train.txt \
  --vali_path=$DATA/vali.txt \
  --test_path=$DATA/test.txt \
  --output_dir=$output_dir \
  --num_features=136 \
  --serank=True \
  --query_label_weight=True
```

tail of the train log is 
```
I0322 18:19:51.462585 140648353728320 evaluation.py:276] Finished evaluation at 2021-03-22-18:19:51
INFO:tensorflow:Saving dict for global step 100000: global_step = 100000, labels_mean = 0.66818184, logits_mean = -0.27165264, loss = 258.99927, metric/arp = 38.71134, metric/ndcg@1 = 0.38938853, metric/ndcg@10 = 0.43623617, metric/ndcg@3 = 0.39571184, metric/ndcg@5 = 0.40858307, metric/ordered_pair_accuracy = 0.6484862
I0322 18:19:51.462839 140648353728320 estimator.py:2066] Saving dict for global step 100000: global_step = 100000, labels_mean = 0.66818184, logits_mean = -0.27165264, loss = 258.99927, metric/arp = 38.71134, metric/ndcg@1 = 0.38938853, metric/ndcg@10 = 0.43623617, metric/ndcg@3 = 0.39571184, metric/ndcg@5 = 0.40858307, metric/ordered_pair_accuracy = 0.6484862
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 100000: outputs/model.ckpt-100000
I0322 18:19:51.463721 140648353728320 estimator.py:2127] Saving 'checkpoint_path' summary for global step 100000: outputs/model.ckpt-100000
```
thx for your help.

**Comment by fangkuann** [comment]
> Hi, thx for your great job. but I have problem about recurrence experiment in web30k. I download the code, and change data to web30k, ndcg@5 of the last ckpt is only 0.409, which is 0.456 in paper.
> I do not change serank.py. and the run_script is
> 
> ```shell
> DATA=/path-to-web30k/Fold1
> 
> output_dir=outputs
> rm -r $output_dir
> mkdir $output_dir
> 
> python serank.py \
>   --train_path=$DATA/train.txt \
>   --vali_path=$DATA/vali.txt \
>   --test_path=$DATA/test.txt \
>   --output_dir=$output_dir \
>   --num_features=136 \
>   --serank=True \
>   --query_label_weight=True
> ```
> 
> tail of the train log is
> 
> ```
> I0322 18:19:51.462585 140648353728320 evaluation.py:276] Finished evaluation at 2021-03-22-18:19:51
> INFO:tensorflow:Saving dict for global step 100000: global_step = 100000, labels_mean = 0.66818184, logits_mean = -0.27165264, loss = 258.99927, metric/arp = 38.71134, metric/ndcg@1 = 0.38938853, metric/ndcg@10 = 0.43623617, metric/ndcg@3 = 0.39571184, metric/ndcg@5 = 0.40858307, metric/ordered_pair_accuracy = 0.6484862
> I0322 18:19:51.462839 140648353728320 estimator.py:2066] Saving dict for global step 100000: global_step = 100000, labels_mean = 0.66818184, logits_mean = -0.27165264, loss = 258.99927, metric/arp = 38.71134, metric/ndcg@1 = 0.38938853, metric/ndcg@10 = 0.43623617, metric/ndcg@3 = 0.39571184, metric/ndcg@5 = 0.40858307, metric/ordered_pair_accuracy = 0.6484862
> INFO:tensorflow:Saving 'checkpoint_path' summary for global step 100000: outputs/model.ckpt-100000
> I0322 18:19:51.463721 140648353728320 estimator.py:2127] Saving 'checkpoint_path' summary for global step 100000: outputs/model.ckpt-100000
> ```
> 
> thx for your help.

Hi,  @stanpcf , thanks for reaching out to us! 

This program is different from the one that we used in the paper experiment. It is simplified and mainly demonstrates the core structure of our proposed method. 
There are a few reasons for the difference in results between this program and paper.

1. The list_size for valid & test should be as large as possible (eg. 1300), because the current limitation of 100 will discard many examples during evaluation. Note that this program runs eval in only one batch, which leads to an OOM error when we set list_size=1300. 
2. We did not implement BestCheckpointExporter in this program. The final output of the test evaluation in this program only evaluates the last checkpoint of training, which may not be the best one.
3. We also met reproducing problems when doing experiments on GSF models a year ago - https://github.com/tensorflow/ranking/issues/148. To our experience, a lot of parameter tuning work is needed to fully reproduce prior works.
4. We will try to provide a new version with comparable results on the paper. stay tuned.

**Comment by wrxDM** [comment]
Hi, @stanpcf, we have just pushed a new version code that supports tfrecord and adds a run_web30k script. The following are the web30k metrics of different models, the performance gain between serank and baseline is nearly consistent with the paper which proves the effectiveness of the serank model. But because of the influence of other factors such as the random seed or training steps we can neither recurrence the actual metric reported on the paper. 
gsf1 metric/ndcg@5 = 0.429
serank metric/ndcg@5 = 0.440
serank-b metric/ndcg@5 = 0.446

**Comment by stanpcf** [comment]
thank you very much :-) @fangkuann @wrxDM

**Comment by stanpcf** [comment]
https://github.com/zhihu/SERank/blob/b6cdef999e5249ddb1a1a3f77ac70f86e96e64be/serank.py#L401-L419
Hi, 
I find that the code use batch_normalization for input normalization(which is common for nn-rank task). but because padding for list, which will cause batch_normalization compute with pad val. 
If there is a batch_normalization with mask which is a better choice for rank task.  

I will be sincerely grateful if any discuss is given.
thanks.
